### DISCUSSION TITLE
Rendering performance of large lists & use of new content-visibility CSS feature

### DISCUSSION THREAD

**Discussion by tobias-klein** [ezra-bible-app/ezra-bible-app#543] [Development]
So, after [reworking the tag list icons](https://github.com/ezra-bible-app/ezra-bible-app/pull/531) the rendering performance became even slower than it was before. The reason was that I added another _wrapper div_ to help with the formatting and that increased the overall number of _divs_ within the container very much. In general, the DOM slows down quite a lot once you go beyond 500 elements in a list. This is an observation that is already years old and until now the only solution seemed to be "lazy loading" / virtual scrolling, etc. - additional effort, complexity and potential for bugs from a programming perspective.

After some research I found a newly available CSS-based solution which is based on optimizations in the browser rendering engines (available since Chrome 85). Welcome `content-visibility: auto`. This new attribute configures the browser to not render certain dom nodes that are off-screen. Once I introduced that attribute on the `.checkbox-tag` class (this is the high-level element within the tag list), the rendering performance dramatically increased, especially on Android, but even noticably on the desktop - I was quite positively suprised about the result. I also added this new CSS attribute to a couple more areas, see 2bd5d9ddfeb63b5a63e6e09a348992e60da599b7, 64d1b7bb25f81644a66fe2e5f8ee82c190aaa259, 7b005f193939f9d686cd2db61b8f4bdc73137598, bbae7e4ba379fbdf0e4b19eda6f44e51b07ed48e. So far, these are the areas where I applied it: Tag assignment list, Tag selection list, Assistant Checkbox (for language list). I think we could also check how it helps to speed up the verse list when applying it to the verse box elements.

For some introduction regarding the concepts related to `content-visibility` have a look here:
https://web.dev/content-visibility
https://developer.mozilla.org/en-US/docs/Web/CSS/content-visibility

**Comment by zhuiks** [comment]
@tobias-klein Thanks for research and optimization. I had looked into this CSS property after rework on module assistant, but I didn't follow up on it with a working solution. So that's very nice that you got it working.
